File: frameNet/src/main/java/org/sqlunet/framenet/sql/FnSemType.java
package org.sqlunet.framenet.sql;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Semantic type
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
class FnSemType
{
	/**
	 * Semantic type id
	 */
	public final long semTypeId;

	/**
	 * Semantic type name
	 */
	public final String semTypeName;

	/**
	 * Semantic type definition
	 */
	public final String semTypeDefinition;

	/**
	 * Constructor
	 *
	 * @param semTypeId         semtype id
	 * @param semTypeName       semtype name
	 * @param semTypeDefinition semtype definition
	 */
	private FnSemType(final long semTypeId, final String semTypeName, final String semTypeDefinition)
	{
		this.semTypeId = semTypeId;
		this.semTypeName = semTypeName;
		this.semTypeDefinition = semTypeDefinition;
	}

	/**
	 * Make semtypes from string
	 *
	 * @param semTypesString (id:def|id:def...)
	 * @return list of semtypes
	 */
	@Nullable
	static public List<FnSemType> make(@Nullable final String semTypesString)
	{
		if (semTypesString == null)
		{
			return null;
		}
		List<FnSemType> result = null;
		final String[] semTypes = semTypesString.split("\\|");
		for (final String semType : semTypes)
		{
			if (result == null)
			{
				result = new ArrayList<>();
			}
			final String[] fields = semType.split(":");
			final long semTypeId = Long.parseLong(fields[0]);
			final String semTypeName = fields[1];
			final String semTypeDefinition = fields[2];
			result.add(new FnSemType(semTypeId, semTypeName, semTypeDefinition));
		}
		return result;
	}
}

File: frameNet/src/main/java/org/sqlunet/framenet/sql/FnSemType.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.framenet.sql

/**
 * Semantic type
 *
 * @param semTypeId         semtype id
 * @param semTypeName       semtype name
 * @param semTypeDefinition semtype definition
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
class FnSemType private constructor(
    val semTypeId: Long,
    val semTypeName: String,
    val semTypeDefinition: String,
) {

    companion object {

        /**
         * Make semtypes from string
         *
         * @param semTypesString (id:def|id:def...)
         * @return list of semtypes
         */
        fun make(semTypesString: String): List<FnSemType>? {
            var result: MutableList<FnSemType>? = null
            val semTypes = semTypesString.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            for (semType in semTypes) {
                if (result == null) {
                    result = ArrayList()
                }
                val fields = semType.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val semTypeId = fields[0].toLong()
                val semTypeName = fields[1]
                val semTypeDefinition = fields[2]
                result.add(FnSemType(semTypeId, semTypeName, semTypeDefinition))
            }
            return result
        }
    }
}
